src/components/orator/js/Orator.js

```diff
@@ -0,0 +1,963 @@
+     *      word: {String} // the text, `word`, parsed from the node. (It may contain only whitespace.)
+     *   }
+     *
+     * @typedef {Object} DomWordMap
+     */
+
+    /**
+     * An array of {DomWordMap} objects detailing the words parsed from the a {DomElement}. This information can be used
+     * to map the synthesized speech to the location of each word in the DOM; which is useful for highlighting words as
+     * they are being read.
+     *
+     * @typedef {Array} DomWordMappings
+     */
+
+    /**
+     * Combines the parsed text into a String.
+     *
+     * @param {DomWordMappings} parsed - An array of {DomWordMap} objects containing the position mappings from a parsed
+     *                                   {DomElement}.
+     *
+     * @return {String} - The parsed text combined into a String.
+     */
+    fluid.orator.domReader.parsedToString = function (parsed) {
+        var words = fluid.transform(parsed, function (block) {
+            return block.word;
+        });
+
+        return words.join("");
+    };
+
+    /**
+     * Parses the DOM element into data points to use for highlighting the text, and queues the text into the self
+     * voicing engine. The parsed data points are added to the component's `parseQueue`
+     *
+     * @param {Component} that - the component
+     * @param {String|jQuery|DomElement} elm - The DOM node to read
+     */
+    fluid.orator.domReader.readFromDOM = function (that, elm) {
+        elm = $(elm);
+
+        // only execute if there are nodes to read from
+        if (elm.length) {
+            var parsedFromElm = that.parser.parse(elm[0]);
+            that.setParseQueue(parsedFromElm);
+            that.queueSpeech(fluid.orator.domReader.parsedToString(parsedFromElm));
+        }
+    };
+
+    /**
+     * Returns the index of the closest data point from the parseQueue based on the boundary provided.
+     *
+     * @param {Component} that - The component
+     * @param {Integer} boundary - The boundary value used to compare against the blockIndex of the parsed data points.
+     *
+     * @return {Integer|undefined} - Will return the index of the closest data point in the parseQueue. If the boundary
+     *                               cannot be located within the parseQueue, `undefined` is returned.
+     */
+    fluid.orator.domReader.getClosestIndex = function (that, boundary) {
+        var parseQueue = that.parseQueue;
+
+        if (!parseQueue.length || !fluid.isValue(boundary) && boundary <= 0) {
+            return undefined;
+        };
+
+        var maxIndex  = Math.max(parseQueue.length - 1, 0);
+        var index = Math.max(Math.min(that.model.parseIndex || 0, maxIndex), 0);
+        var maxBoundary = parseQueue[maxIndex].blockIndex + parseQueue[maxIndex].word.length;
+
+        if (boundary > maxBoundary || boundary < 0) {
+            return undefined;
+        }
+
+        while (index >= 0) {
```

Comment:
Thanks, I think the workflow of this function is much more comprehensible now. Also it will be hugely more efficient :)